overlord/snapstate/snapmgr.go

```diff
@@ -389,24 +399,24 @@ func (m *SnapManager) doPrepareSnap(t *state.Task, _ *tomb.Tomb) error {
 			panic("internal error: invalid local revision built: " + revision.String())
 		}
 		snapst.LocalRevision = revision
-		ss.Revision = revision
-		snapst.Candidate = &snap.SideInfo{Revision: ss.Revision}
+		ss.SideInfo.Revision = revision
+		snapst.Candidate = &snap.SideInfo{Revision: ss.Revision()}
 	} else {
 		for _, si := range snapst.Sequence {
-			if si.Revision == ss.Revision {
+			if si.Revision == ss.Revision() {
 				snapst.Candidate = si
 				break
 			}
 		}
 	}
 
 	if snapst.Candidate == nil {
```

Comment:
if both path of the "if" would end with snapst.Candidate = ss.SideInfo, that could be pulled out and this check wouldn't be needed anymore